coap/server.go

```diff
@@ -1,30 +1,129 @@
-package adapter
+package coap
 
 import (
-	"log"
+	"fmt"
 	"net"
+	"strings"
+	"time"
 
-	"github.com/dereulenspiegel/coap-mux"
-	"github.com/dustin/go-coap"
+	mux "github.com/dereulenspiegel/coap-mux"
+	manager "github.com/mainflux/mainflux/manager/client"
+
+	gocoap "github.com/dustin/go-coap"
 )
 
-func notFoundHandler(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
-	log.Printf("Got message in notFoundHandler: path=%q: %#v from %v", m.Path(), m, a)
-	if m.IsConfirmable() {
-		return &coap.Message{
-			Type: coap.Acknowledgement,
-			Code: coap.NotFound,
+var auth manager.ManagerClient
+
+func authKey(opt interface{}) (string, error) {
+	if opt == nil {
+		return "", errBadRequest
+	}
+	val, ok := opt.(string)
+	if !ok {
+		return "", errBadRequest
+	}
+	arr := strings.Split(val, "=")
+	if len(arr) != 2 || strings.ToLower(arr[0]) != "key" {
+		return "", errBadOption
+	}
+	return arr[1], nil
+}
+
+// Authorize method is used to authorize request.
+func Authorize(msg *gocoap.Message, res *gocoap.Message, cid string) (publisher string, err error) {
+	// Device Key is passed as Uri-Query parameter, which option ID is 15 (0xf).
+	key, err := authKey(msg.Option(gocoap.URIQuery))
+	if err != nil {
+		switch err {
+		case errBadOption:
+			res.Code = gocoap.BadOption
+		case errBadRequest:
+			res.Code = gocoap.BadRequest
+		}
+		return
+	}
+
+	publisher, err = auth.CanAccess(cid, key)
+	if err != nil {
+		switch err {
+		case manager.ErrServiceUnreachable:
+			res.Code = gocoap.ServiceUnavailable
+		default:
+			res.Code = gocoap.Unauthorized
+		}
+	}
+	return
+}
+
+func serve(svc Service, conn *net.UDPConn, data []byte, addr *net.UDPAddr, rh gocoap.Handler) {
+	msg, err := gocoap.ParseMessage(data)
+	if err != nil {
+		return
+	}
+	res := &gocoap.Message{
+		Type:      gocoap.NonConfirmable,
+		Code:      gocoap.Content,
+		MessageID: msg.MessageID,
+		Token:     msg.Token,
+		Payload:   []byte{},
+	}
+	switch msg.Type {
+	case gocoap.Reset:
+		cid := mux.Var(&msg, "id")
+		res.Type = gocoap.Acknowledgement
+		publisher, err := Authorize(&msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+		} else {
```

Comment:
Avoid `else` by using `break`.

Reply:
Using `break` is probably faster, but since this is typical `if-else` case, I'd prefer to keep it this way.